File: main/counting-valleys/counting-valleys.py
#!/usr/bin/env python3

def count_valleys(trek):
    level = count = 0
    
    for step in trek:
        if step == 'U':
            level += 1
            if level == 0:
                count += 1
        elif step == 'D':
            level -= 1
        else:
            raise ValueError('unrecognized step')
    
    return count

length = int(input())
trek = input()

if len(trek) != length:
    raise ValueError('wrong number of steps')

print(count_valleys(trek))

File: main/counting-valleys/counting-valleys.pl
#!/usr/bin/env perl

use strict;
use warnings;
use Carp qw(croak);

sub count_valleys {
    my ($trek) = @_;

    my $level = 0;
    my $count = 0;

    for my $step (split //ms, $trek) {
        if ($step eq 'U') {
            if (++$level == 0) { ++$count }
        } elsif ($step eq 'D') {
            --$level;
        } else {
            croak 'unrecognized step';
        }
    }

    return $count;
}

my $len = <>;
chomp(my $trek = <>);
length $trek == $len or croak 'wrong number of steps';

print count_valleys($trek), "\n";
